File: Braintree/src/test/java/com/braintreepayments/api/models/ThreeDSecureAdditionalInformationTest.java
package com.braintreepayments.api.models;

import android.os.Parcel;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
public class ThreeDSecureAdditionalInformationTest {

    @Test
    public void constructsCorrectly() {
        ThreeDSecurePostalAddress postalAddress = new ThreeDSecurePostalAddress()
                .streetAddress("123 Fake St.")
                .extendedAddress("Apt. 3")
                .locality("Oakland")
                .region("CA")
                .postalCode("94602")
                .countryCodeAlpha2("US")
                .firstName("John")
                .lastName("Fakerson");

        ThreeDSecureAdditionalInformation additionalInformation = new ThreeDSecureAdditionalInformation()
                .billingAddress(postalAddress)
                .billingPhoneNumber("billing-phone-number")
                .billingGivenName("billing-given-name")
                .billingSurname("billing-surname")
                .email("email")
                .shippingMethod("shipping-method");

        assertEquals("billing-phone-number", additionalInformation.getBillingPhoneNumber());
        assertEquals("billing-given-name", additionalInformation.getBillingGivenName());
        assertEquals("billing-surname", additionalInformation.getBillingSurname());
        assertEquals("email", additionalInformation.getEmail());
        assertEquals("shipping-method", additionalInformation.getShippingMethod());
        assertEquals(postalAddress, additionalInformation.getBillingAddress());
    }

    @Test
    public void testWriteToParcel_serializesCorrectly() {
        ThreeDSecurePostalAddress postalAddress = new ThreeDSecurePostalAddress()
                .streetAddress("street-address")
                .extendedAddress("extended-address")
                .countryCodeAlpha2("country-code");

        ThreeDSecureAdditionalInformation preSerialized = new ThreeDSecureAdditionalInformation()
                .billingAddress(postalAddress)
                .billingPhoneNumber("billing-phone-number")
                .billingGivenName("billing-given-name")
                .billingSurname("billing-surname")
                .email("email")
                .shippingMethod("shipping-method")
                .shippingMethodIndicator("shipping-method-indicator")
                .productCode("productCode")
                .deliveryTimeframe("deliveryTimeframe")
                .deliveryEmail("deliveryEmail")
                .reorderIndicator("reorderIndicator")
                .preorderIndicator("preorderIndicator")
                .preorderDate("preorderDate")
                .giftCardAmount("giftCardAmount")
                .giftCardCurrencyCode("giftCardCurrencyCode")
                .giftCardCount("giftCardCount")
                .accountAgeIndicator("accountAgeIndicator")
                .accountCreateDate("accountCreateDate")
                .accountChangeIndicator("accountChangeIndicator")
                .accountChangeDate("accountChangeDate")
                .accountPwdChangeIndicator("accountPwdChangeIndicator")
                .accountPwdChangeDate("accountPwdChangeDate")
                .shippingAddressUsageIndicator("shippingAddressUsageIndicator")
                .shippingAddressUsageDate("shippingAddressUsageDate")
                .transactionCountDay("transactionCountDay")
                .transactionCountYear("transactionCountYear")
                .addCardAttempts("addCardAttempts")
                .accountPurchases("accountPurchases")
                .fraudActivity("fraudActivity")
                .shippingNameIndicator("shippingNameIndicator")
                .paymentAccountIndicator("paymentAccountIndicator")
                .paymentAccountAge("paymentAccountAge")
                .addressMatch("addressMatch")
                .accountId("accountId")
                .ipAddress("ipAddress")
                .orderDescription("orderDescription")
                .taxAmount("taxAmount")
                .userAgent("userAgent")
                .authenticationIndicator("authenticationIndicator")
                .installment("installment")
                .purchaseDate("purchaseDate")
                .recurringEnd("recurringEnd")
                .recurringFrequency("recurringFrequency");

        Parcel parcel = Parcel.obtain();
        preSerialized.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);

        ThreeDSecureAdditionalInformation postSerialized = ThreeDSecureAdditionalInformation.CREATOR.createFromParcel(parcel);

        assertNotNull(postSerialized);
        assertEquals("billing-phone-number", postSerialized.getBillingPhoneNumber());
        assertEquals("billing-given-name", postSerialized.getBillingGivenName());
        assertEquals("billing-surname", postSerialized.getBillingSurname());
        assertEquals("email", postSerialized.getEmail());
        assertEquals("shipping-method", postSerialized.getShippingMethod());
        assertEquals("country-code", postSerialized.getBillingAddress().getCountryCodeAlpha2());
        assertEquals("street-address", postSerialized.getBillingAddress().getStreetAddress());
        assertEquals("extended-address", postSerialized.getBillingAddress().getExtendedAddress());
        assertEquals("shipping-method", postSerialized.getShippingMethod());
        assertEquals("shipping-method-indicator", postSerialized.getShippingMethodIndicator());
        assertEquals("productCode", postSerialized.getProductCode());
        assertEquals("deliveryTimeframe", postSerialized.getDeliveryTimeframe());
        assertEquals("deliveryEmail", postSerialized.getDeliveryEmail());
        assertEquals("reorderIndicator", postSerialized.getReorderIndicator());
        assertEquals("preorderIndicator", postSerialized.getPreorderIndicator());
        assertEquals("preorderDate", postSerialized.getPreorderDate());
        assertEquals("giftCardAmount", postSerialized.getGiftCardAmount());
        assertEquals("giftCardCurrencyCode", postSerialized.getGiftCardCurrencyCode());
        assertEquals("giftCardCount", postSerialized.getGiftCardCount());
        assertEquals("accountAgeIndicator", postSerialized.getAccountAgeIndicator());
        assertEquals("accountCreateDate", postSerialized.getAccountCreateDate());
        assertEquals("accountChangeIndicator", postSerialized.getAccountChangeIndicator());
        assertEquals("accountChangeDate", postSerialized.getAccountChangeDate());
        assertEquals("accountPwdChangeIndicator", postSerialized.getAccountPwdChangeIndicator());
        assertEquals("accountPwdChangeDate", postSerialized.getAccountPwdChangeDate());
        assertEquals("shippingAddressUsageIndicator", postSerialized.getShippingAddressUsageIndicator());
        assertEquals("shippingAddressUsageDate", postSerialized.getShippingAddressUsageDate());
        assertEquals("transactionCountDay", postSerialized.getTransactionCountDay());
        assertEquals("transactionCountYear", postSerialized.getTransactionCountYear());
        assertEquals("addCardAttempts", postSerialized.getAddCardAttempts());
        assertEquals("accountPurchases", postSerialized.getAccountPurchases());
        assertEquals("fraudActivity", postSerialized.getFraudActivity());
        assertEquals("shippingNameIndicator", postSerialized.getShippingNameIndicator());
        assertEquals("paymentAccountIndicator", postSerialized.getPaymentAccountIdicator());
        assertEquals("paymentAccountAge", postSerialized.getPaymentAccountAge());
        assertEquals("addressMatch", postSerialized.getAddressMatch());
        assertEquals("accountId", postSerialized.getAccountId());
        assertEquals("ipAddress", postSerialized.getIpAddress());
        assertEquals("orderDescription", postSerialized.getOrderDescription());
        assertEquals("taxAmount", postSerialized.getTaxAmount());
        assertEquals("userAgent", postSerialized.getUserAgent());
        assertEquals("authenticationIndicator", postSerialized.getAuthenticationIndicator());
        assertEquals("installment", postSerialized.getInstallment());
        assertEquals("purchaseDate", postSerialized.getPurchaseDate());
        assertEquals("recurringEnd", postSerialized.getRecurringEnd());
        assertEquals("recurringFrequency", postSerialized.getRecurringFrequency());
    }

    @Test
    public void testToJson_buildsAllParameters() throws JSONException {
        ThreeDSecurePostalAddress postalAddress = new ThreeDSecurePostalAddress()
                .streetAddress("street-address")
                .extendedAddress("extended-address")
                .locality("locality");

        ThreeDSecureAdditionalInformation additionalInformation = new ThreeDSecureAdditionalInformation()
                .billingAddress(postalAddress)
                .billingPhoneNumber("billing-phone-number")
                .billingGivenName("billing-given-name")
                .billingSurname("billing-surname")
                .email("email")
                .shippingMethod("shipping-method");

        JSONObject jsonParams = additionalInformation.toJson();

        assertEquals("billing-given-name", jsonParams.get("firstName"));
        assertEquals("billing-surname", jsonParams.get("lastName"));
        assertEquals("billing-phone-number", jsonParams.get("mobilePhoneNumber"));
        assertEquals("email", jsonParams.get("email"));
        assertEquals("shipping-method", jsonParams.get("shippingMethod"));
        assertEquals("street-address", jsonParams.get("line1"));
        assertEquals("extended-address", jsonParams.get("line2"));
        assertEquals("locality", jsonParams.get("city"));
    }

    @Test
    public void testToJson_buildsPartialParameters() throws JSONException {
        ThreeDSecurePostalAddress postalAddress = new ThreeDSecurePostalAddress()
                .streetAddress("street-address")
                .extendedAddress("extended-address")
                .locality("locality");

        ThreeDSecureAdditionalInformation additionalInformation = new ThreeDSecureAdditionalInformation()
                .billingAddress(postalAddress)
                .billingSurname("billing-surname")
                .email("email")
                .shippingMethod("shipping-method")
                .shippingMethodIndicator("shipping-method-indicator")
                .productCode("productCode")
                .deliveryTimeframe("deliveryTimeframe")
                .deliveryEmail("deliveryEmail")
                .reorderIndicator("reorderIndicator")
                .preorderIndicator("preorderIndicator")
                .preorderDate("preorderDate")
                .giftCardAmount("giftCardAmount")
                .giftCardCurrencyCode("giftCardCurrencyCode")
                .giftCardCount("giftCardCount")
                .accountAgeIndicator("accountAgeIndicator")
                .accountCreateDate("accountCreateDate")
                .accountChangeIndicator("accountChangeIndicator")
                .accountChangeDate("accountChangeDate")
                .accountPwdChangeIndicator("accountPwdChangeIndicator")
                .accountPwdChangeDate("accountPwdChangeDate")
                .shippingAddressUsageIndicator("shippingAddressUsageIndicator")
                .shippingAddressUsageDate("shippingAddressUsageDate")
                .transactionCountDay("transactionCountDay")
                .transactionCountYear("transactionCountYear")
                .addCardAttempts("addCardAttempts")
                .accountPurchases("accountPurchases")
                .fraudActivity("fraudActivity")
                .shippingNameIndicator("shippingNameIndicator")
                .paymentAccountIndicator("paymentAccountIndicator")
                .paymentAccountAge("paymentAccountAge")
                .addressMatch("addressMatch")
                .accountId("accountId")
                .ipAddress("ipAddress")
                .orderDescription("orderDescription")
                .taxAmount("taxAmount")
                .userAgent("userAgent")
                .authenticationIndicator("authenticationIndicator")
                .installment("installment")
                .purchaseDate("purchaseDate")
                .recurringEnd("recurringEnd")
                .recurringFrequency("recurringFrequency");

        JSONObject jsonParams = additionalInformation.toJson();

        assertTrue(jsonParams.isNull("firstName"));
        assertEquals("billing-surname", jsonParams.get("lastName"));
        assertTrue(jsonParams.isNull("mobilePhoneNumber"));
        assertEquals("street-address", jsonParams.get("line1"));
        assertEquals("extended-address", jsonParams.get("line2"));
        assertTrue(jsonParams.isNull("postalCode"));
        assertTrue(jsonParams.isNull("state"));
        assertTrue(jsonParams.isNull("countryCode"));
        assertEquals("shipping-method-indicator", jsonParams.getString("shippingMethodIndicator"));
        assertEquals("productCode", jsonParams.getString("productCode"));
        assertEquals("deliveryTimeframe", jsonParams.getString("deliveryTimeframe"));
        assertEquals("deliveryEmail", jsonParams.getString("deliveryEmail"));
        assertEquals("reorderIndicator", jsonParams.getString("reorderIndicator"));
        assertEquals("preorderIndicator", jsonParams.getString("preorderIndicator"));
        assertEquals("preorderDate", jsonParams.getString("preorderDate"));
        assertEquals("giftCardAmount", jsonParams.getString("giftCardAmount"));
        assertEquals("giftCardCurrencyCode", jsonParams.getString("giftCardCurrencyCode"));
        assertEquals("giftCardCount", jsonParams.getString("giftCardCount"));
        assertEquals("accountAgeIndicator", jsonParams.getString("accountAgeIndicator"));
        assertEquals("accountCreateDate", jsonParams.getString("accountCreateDate"));
        assertEquals("accountChangeIndicator", jsonParams.getString("accountChangeIndicator"));
        assertEquals("accountChangeDate", jsonParams.getString("accountChangeDate"));
        assertEquals("accountPwdChangeIndicator", jsonParams.getString("accountPwdChangeIndicator"));
        assertEquals("accountPwdChangeDate", jsonParams.getString("accountPwdChangeDate"));
        assertEquals("shippingAddressUsageIndicator", jsonParams.getString("shippingAddressUsageIndicator"));
        assertEquals("shippingAddressUsageDate", jsonParams.getString("shippingAddressUsageDate"));
        assertEquals("transactionCountDay", jsonParams.getString("transactionCountDay"));
        assertEquals("transactionCountYear", jsonParams.getString("transactionCountYear"));
        assertEquals("addCardAttempts", jsonParams.getString("addCardAttempts"));
        assertEquals("accountPurchases", jsonParams.getString("accountPurchases"));
        assertEquals("fraudActivity", jsonParams.getString("fraudActivity"));
        assertEquals("shippingNameIndicator", jsonParams.getString("shippingNameIndicator"));
        assertEquals("paymentAccountIndicator", jsonParams.getString("paymentAccountIndicator"));
        assertEquals("paymentAccountAge", jsonParams.getString("paymentAccountAge"));
        assertEquals("addressMatch", jsonParams.getString("addressMatch"));
        assertEquals("accountId", jsonParams.getString("accountId"));
        assertEquals("ipAddress", jsonParams.getString("ipAddress"));
        assertEquals("orderDescription", jsonParams.getString("orderDescription"));
        assertEquals("taxAmount", jsonParams.getString("taxAmount"));
        assertEquals("userAgent", jsonParams.getString("userAgent"));
        assertEquals("authenticationIndicator", jsonParams.getString("authenticationIndicator"));
        assertEquals("installment", jsonParams.getString("installment"));
        assertEquals("purchaseDate", jsonParams.getString("purchaseDate"));
        assertEquals("recurringEnd", jsonParams.getString("recurringEnd"));
        assertEquals("recurringFrequency", jsonParams.getString("recurringFrequency"));
    }

    @Test
    public void testToJson_buildsEmptyParameters() {
        ThreeDSecureAdditionalInformation additionalInformation = new ThreeDSecureAdditionalInformation();

        JSONObject jsonParams = additionalInformation.toJson();

        assertTrue(jsonParams.isNull("billingAddress"));
        assertTrue(jsonParams.isNull("firstName"));
        assertTrue(jsonParams.isNull("lastName"));
        assertTrue(jsonParams.isNull("shippingMethod"));
        assertTrue(jsonParams.isNull("mobilePhoneNumber"));
        assertTrue(jsonParams.isNull("shippingMethodIndicator"));
        assertTrue(jsonParams.isNull("productCode"));
        assertTrue(jsonParams.isNull("deliveryTimeframe"));
        assertTrue(jsonParams.isNull("deliveryEmail"));
        assertTrue(jsonParams.isNull("reorderIndicator"));
        assertTrue(jsonParams.isNull("preorderIndicator"));
        assertTrue(jsonParams.isNull("preorderDate"));
        assertTrue(jsonParams.isNull("giftCardAmount"));
        assertTrue(jsonParams.isNull("giftCardCurrencyCode"));
        assertTrue(jsonParams.isNull("giftCardCount"));
        assertTrue(jsonParams.isNull("accountAgeIndicator"));
        assertTrue(jsonParams.isNull("accountCreateDate"));
        assertTrue(jsonParams.isNull("accountChangeIndicator"));
        assertTrue(jsonParams.isNull("accountChangeDate"));
        assertTrue(jsonParams.isNull("accountPwdChangeIndicator"));
        assertTrue(jsonParams.isNull("accountPwdChangeDate"));
        assertTrue(jsonParams.isNull("shippingAddressUsageIndicator"));
        assertTrue(jsonParams.isNull("shippingAddressUsageDate"));
        assertTrue(jsonParams.isNull("transactionCountDay"));
        assertTrue(jsonParams.isNull("transactionCountYear"));
        assertTrue(jsonParams.isNull("addCardAttempts"));
        assertTrue(jsonParams.isNull("accountPurchases"));
        assertTrue(jsonParams.isNull("fraudActivity"));
        assertTrue(jsonParams.isNull("shippingNameIndicator"));
        assertTrue(jsonParams.isNull("paymentAccountIndicator"));
        assertTrue(jsonParams.isNull("paymentAccountAge"));
        assertTrue(jsonParams.isNull("addressMatch"));
        assertTrue(jsonParams.isNull("accountId"));
        assertTrue(jsonParams.isNull("ipAddress"));
        assertTrue(jsonParams.isNull("orderDescription"));
        assertTrue(jsonParams.isNull("taxAmount"));
        assertTrue(jsonParams.isNull("userAgent"));
        assertTrue(jsonParams.isNull("authenticationIndicator"));
        assertTrue(jsonParams.isNull("installment"));
        assertTrue(jsonParams.isNull("purchaseDate"));
        assertTrue(jsonParams.isNull("recurringEnd"));
        assertTrue(jsonParams.isNull("recurringFrequency"));
    }
}

File: ThreeDSecure/src/androidTest/java/com/braintreepayments/api/threedsecure/ThreeDSecureAdditionalInformationTest.kt
package com.braintreepayments.api.threedsecure

import android.os.Parcel
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
import kotlinx.parcelize.parcelableCreator
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4ClassRunner::class)
class ThreeDSecureAdditionalInformationTest {

    @Suppress("LongMethod")
    @Test
    fun parcels_withAllFieldsPopulated() {
        val shippingAddress = ThreeDSecurePostalAddress(
            givenName = "Joe",
            surname = "Guy",
            streetAddress = "555 Smith St.",
            extendedAddress = "Unit 1",
            line3 = "Building A",
            locality = "Oakland",
            region = "CA",
            postalCode = "12345",
            countryCodeAlpha2 = "US",
            phoneNumber = "5551234567"
        )

        val original = ThreeDSecureAdditionalInformation(
            shippingAddress = shippingAddress,
            shippingMethodIndicator = "01",
            productCode = "AIR",
            deliveryTimeframe = "02",
            deliveryEmail = "test@example.com",
            reorderIndicator = "01",
            preorderIndicator = "02",
            preorderDate = "20250101",
            giftCardAmount = "100",
            giftCardCurrencyCode = "USD",
            giftCardCount = "2",
            accountAgeIndicator = "05",
            accountCreateDate = "20200101",
            accountChangeIndicator = "04",
            accountChangeDate = "20240601",
            accountPwdChangeIndicator = "03",
            accountPwdChangeDate = "20240501",
            shippingAddressUsageIndicator = "04",
            shippingAddressUsageDate = "20230101",
            transactionCountDay = "3",
            transactionCountYear = "50",
            addCardAttempts = "1",
            accountPurchases = "20",
            fraudActivity = "01",
            shippingNameIndicator = "01",
            paymentAccountIndicator = "05",
            paymentAccountAge = "20190601",
            addressMatch = "Y",
            accountId = "user-123",
            ipAddress = "192.168.1.1",
            orderDescription = "Test order",
            taxAmount = "1000",
            userAgent = "Mozilla/5.0",
            authenticationIndicator = "02",
            installment = "3",
            purchaseDate = "20250101120000",
            recurringEnd = "20261231",
            recurringFrequency = "28",
            sdkMaxTimeout = "05",
            workPhoneNumber = "5559876543"
        )

        val parcel = Parcel.obtain()
        original.writeToParcel(parcel, 0)
        parcel.setDataPosition(0)

        val restored = parcelableCreator<ThreeDSecureAdditionalInformation>().createFromParcel(parcel)
        parcel.recycle()

        assertEquals(original.shippingAddress?.givenName, restored.shippingAddress?.givenName)
        assertEquals(original.shippingAddress?.surname, restored.shippingAddress?.surname)
        assertEquals(original.shippingAddress?.streetAddress, restored.shippingAddress?.streetAddress)
        assertEquals(original.shippingAddress?.locality, restored.shippingAddress?.locality)
        assertEquals(original.shippingAddress?.postalCode, restored.shippingAddress?.postalCode)
        assertEquals(original.shippingMethodIndicator, restored.shippingMethodIndicator)
        assertEquals(original.productCode, restored.productCode)
        assertEquals(original.deliveryTimeframe, restored.deliveryTimeframe)
        assertEquals(original.deliveryEmail, restored.deliveryEmail)
        assertEquals(original.reorderIndicator, restored.reorderIndicator)
        assertEquals(original.preorderIndicator, restored.preorderIndicator)
        assertEquals(original.preorderDate, restored.preorderDate)
        assertEquals(original.giftCardAmount, restored.giftCardAmount)
        assertEquals(original.giftCardCurrencyCode, restored.giftCardCurrencyCode)
        assertEquals(original.giftCardCount, restored.giftCardCount)
        assertEquals(original.accountAgeIndicator, restored.accountAgeIndicator)
        assertEquals(original.accountCreateDate, restored.accountCreateDate)
        assertEquals(original.accountChangeIndicator, restored.accountChangeIndicator)
        assertEquals(original.accountChangeDate, restored.accountChangeDate)
        assertEquals(original.accountPwdChangeIndicator, restored.accountPwdChangeIndicator)
        assertEquals(original.accountPwdChangeDate, restored.accountPwdChangeDate)
        assertEquals(original.shippingAddressUsageIndicator, restored.shippingAddressUsageIndicator)
        assertEquals(original.shippingAddressUsageDate, restored.shippingAddressUsageDate)
        assertEquals(original.transactionCountDay, restored.transactionCountDay)
        assertEquals(original.transactionCountYear, restored.transactionCountYear)
        assertEquals(original.addCardAttempts, restored.addCardAttempts)
        assertEquals(original.accountPurchases, restored.accountPurchases)
        assertEquals(original.fraudActivity, restored.fraudActivity)
        assertEquals(original.shippingNameIndicator, restored.shippingNameIndicator)
        assertEquals(original.paymentAccountIndicator, restored.paymentAccountIndicator)
        assertEquals(original.paymentAccountAge, restored.paymentAccountAge)
        assertEquals(original.addressMatch, restored.addressMatch)
        assertEquals(original.accountId, restored.accountId)
        assertEquals(original.ipAddress, restored.ipAddress)
        assertEquals(original.orderDescription, restored.orderDescription)
        assertEquals(original.taxAmount, restored.taxAmount)
        assertEquals(original.userAgent, restored.userAgent)
        assertEquals(original.authenticationIndicator, restored.authenticationIndicator)
        assertEquals(original.installment, restored.installment)
        assertEquals(original.purchaseDate, restored.purchaseDate)
        assertEquals(original.recurringEnd, restored.recurringEnd)
        assertEquals(original.recurringFrequency, restored.recurringFrequency)
        assertEquals(original.sdkMaxTimeout, restored.sdkMaxTimeout)
        assertEquals(original.workPhoneNumber, restored.workPhoneNumber)
    }

    @Test
    fun parcels_withDefaultValues() {
        val original = ThreeDSecureAdditionalInformation()

        val parcel = Parcel.obtain()
        original.writeToParcel(parcel, 0)
        parcel.setDataPosition(0)

        val restored = parcelableCreator<ThreeDSecureAdditionalInformation>().createFromParcel(parcel)
        parcel.recycle()

        assertNull(restored.shippingAddress)
        assertNull(restored.shippingMethodIndicator)
        assertNull(restored.productCode)
        assertNull(restored.deliveryTimeframe)
        assertNull(restored.deliveryEmail)
        assertNull(restored.accountId)
        assertNull(restored.ipAddress)
        assertNull(restored.workPhoneNumber)
    }

    @Test
    fun parcels_withNestedShippingAddress() {
        val shippingAddress = ThreeDSecurePostalAddress(
            givenName = "Jane",
            surname = "Doe",
            streetAddress = "123 Main St",
            extendedAddress = "Suite 200",
            line3 = "Floor 3",
            locality = "Chicago",
            region = "IL",
            postalCode = "60601",
            countryCodeAlpha2 = "US",
            phoneNumber = "3125551234"
        )
        val original = ThreeDSecureAdditionalInformation(shippingAddress = shippingAddress)

        val parcel = Parcel.obtain()
        original.writeToParcel(parcel, 0)
        parcel.setDataPosition(0)

        val restored = parcelableCreator<ThreeDSecureAdditionalInformation>().createFromParcel(parcel)
        parcel.recycle()

        assertEquals("Jane", restored.shippingAddress?.givenName)
        assertEquals("Doe", restored.shippingAddress?.surname)
        assertEquals("123 Main St", restored.shippingAddress?.streetAddress)
        assertEquals("Suite 200", restored.shippingAddress?.extendedAddress)
        assertEquals("Floor 3", restored.shippingAddress?.line3)
        assertEquals("Chicago", restored.shippingAddress?.locality)
        assertEquals("IL", restored.shippingAddress?.region)
        assertEquals("60601", restored.shippingAddress?.postalCode)
        assertEquals("US", restored.shippingAddress?.countryCodeAlpha2)
        assertEquals("3125551234", restored.shippingAddress?.phoneNumber)
    }

    @Test
    fun toJson_withAllFieldsPopulated() {
        val shippingAddress = ThreeDSecurePostalAddress(
            givenName = "Joe",
            surname = "Guy",
            streetAddress = "555 Smith St.",
            locality = "Oakland",
            region = "CA",
            postalCode = "12345",
            countryCodeAlpha2 = "US",
            phoneNumber = "5551234567"
        )

        val additionalInfo = ThreeDSecureAdditionalInformation(
            shippingAddress = shippingAddress,
            shippingMethodIndicator = "01",
            productCode = "AIR",
            deliveryTimeframe = "02",
            deliveryEmail = "test@example.com",
            reorderIndicator = "01",
            preorderIndicator = "02",
            preorderDate = "20250101",
            giftCardAmount = "100",
            giftCardCurrencyCode = "USD",
            giftCardCount = "2",
            accountAgeIndicator = "05",
            accountCreateDate = "20200101",
            ipAddress = "192.168.1.1",
            orderDescription = "Test order",
            taxAmount = "1000",
            workPhoneNumber = "5559876543"
        )

        val json = additionalInfo.toJson()

        assertEquals("Joe", json.getString("shipping_given_name"))
        assertEquals("Guy", json.getString("shipping_surname"))
        assertEquals("555 Smith St.", json.getString("shipping_line1"))
        assertEquals("Oakland", json.getString("shipping_city"))
        assertEquals("CA", json.getString("shipping_state"))
        assertEquals("12345", json.getString("shipping_postal_code"))
        assertEquals("US", json.getString("shipping_country_code"))
        assertEquals("5551234567", json.getString("shipping_phone"))
        assertEquals("01", json.getString("shipping_method_indicator"))
        assertEquals("AIR", json.getString("product_code"))
        assertEquals("02", json.getString("delivery_timeframe"))
        assertEquals("test@example.com", json.getString("delivery_email"))
        assertEquals("01", json.getString("reorder_indicator"))
        assertEquals("02", json.getString("preorder_indicator"))
        assertEquals("20250101", json.getString("preorder_date"))
        assertEquals("100", json.getString("gift_card_amount"))
        assertEquals("USD", json.getString("gift_card_currency_code"))
        assertEquals("2", json.getString("gift_card_count"))
        assertEquals("05", json.getString("account_age_indicator"))
        assertEquals("20200101", json.getString("account_create_date"))
        assertEquals("192.168.1.1", json.getString("ip_address"))
        assertEquals("Test order", json.getString("order_description"))
        assertEquals("1000", json.getString("tax_amount"))
        assertEquals("5559876543", json.getString("work_phone_number"))
    }

    @Test
    fun toJson_withEmptyFields() {
        val additionalInfo = ThreeDSecureAdditionalInformation()

        val json = additionalInfo.toJson()

        assertEquals(0, json.length())
    }
}
